src/poly/cse.rs

```diff
@@ -1,134 +1,121 @@
-use rand_chacha::{rand_core::SeedableRng, ChaCha20Rng};
-
 use crate::field::Field;
 
-use std::{
-    collections::HashMap,
-    fmt::Debug,
-    hash::Hash,
-    rc::{Rc, Weak},
-};
-
-use super::{Expr, HashResult, VarAssignments};
-
-/// Common Subexpression Elimination - takes a collection of expr
-pub fn cse<F: Field + Hash, V: Debug + Clone + Eq + Hash>(
-    exprs: Vec<Expr<F, V, ()>>,
-    queriables: &Vec<V>,
-) -> Vec<Rc<Expr<F, V, HashResult>>> {
-    // generate random point in the field set
-    let assignments = generate_random_assignment(queriables);
-
-    // hash table with the hash of the expression as the key
-    // and the weak reference to the expression as the value
-    let mut seen_hashes: HashMap<u64, Weak<Expr<F, V, HashResult>>> = HashMap::new();
-
-    let mut result = Vec::new();
-
-    for expr in exprs {
-        let hashed_expr = Rc::new(expr.hash(&assignments));
-        let hash = hashed_expr.meta().hash;
+use std::{fmt::Debug, hash::Hash};
+
+use super::{ConstrDecomp, Expr, HashResult, SignalFactory, VarAssignments};
+
+pub fn replace_common_subexprs<
+    F: Field + Hash,
+    V: Clone + Eq + PartialEq + Hash + Debug,
+    SF: SignalFactory<V>,
+>(
+    constr: Expr<F, V, ()>,
+    common_ses: &[Expr<F, V, HashResult>],
+    assignments: &VarAssignments<F, V>,
+    signal_factory: &mut SF,
+) -> (Expr<F, V, ()>, ConstrDecomp<F, V>) {
+    let mut decomp = ConstrDecomp::default();
+    let expr =
+        replace_common_subexprs_rec(&mut decomp, constr, common_ses, assignments, signal_factory);
+
+    (expr, decomp)
+}
 
-        // if the hash is already in the hash table,
-        // push the existing expression to the result
-        if let Some(existing_weak) = seen_hashes.get(&hash) {
-            if let Some(existing_expr) = existing_weak.upgrade() {
-                result.push(existing_expr);
-                continue;
+fn replace_common_subexprs_rec<
+    F: Field + Hash,
+    V: Clone + Eq + PartialEq + Hash + Debug,
+    SF: SignalFactory<V>,
+>(
+    decomp: &mut ConstrDecomp<F, V>,
+    constr: Expr<F, V, ()>,
+    common_ses: &[Expr<F, V, HashResult>],
+    assignments: &VarAssignments<F, V>,
+    signal_factory: &mut SF,
+) -> Expr<F, V, ()> {
```

Comment:
@rutefig why constr doesn't have HashResult? We should have done it at this point for all of them

Reply:
This bottom up, what component you need?

So I would do a function that subsitute one common_se into a constr, instead of an array, make the loop outside.

i would have something like: (decomp: &mut ConstrDecomp<F, V>, constr: Expr<F, V, HashResult>, common_se: expr<F, V, HashResult>, assignments: &VarAssignments<F, V>, signal_factory: &mut SF)

Reply:
Also, a better way to keep track of the signals that we have created to represent a common subexpression. Perhaps adding method to ConstrDecomp?

Reply:
So I think the loop is better to be outside because this way I will check from higher degree to lower naturally and replacing as soon as it finds it. And I followed the same pattern as the other optimisations, this way the code is more easy to understand I think...

Reply:
I don't know which loop you mean. I think the function should substitute one sub expression at a time, to control which subsexpressions get substituted first from outside

Reply:
And that's how is being done' I'm just passing a reference to the list of the already found common ses and then I'm checking for each expression it it is a common one or if it has some common subexpression. I just compute the common ses all before replacing it because it gives more control on which should be replaced or not